### DISCUSSION TITLE
Watching for hash changes

### DISCUSSION THREAD

**Discussion by musjj** [watchexec/watchexec#807] [Questions & Help] [answered]
While waiting for https://github.com/watchexec/watchexec/pull/571 to be merged, are there any alternative solutions for watching for file hash changes?

I tried exporting shell functions and calling it from `watchexec`, but the problem is that global variables (that stores the previous hash) are not retained between calls. Are there any better ways of doing this?

**Comment by passcod** [comment]
You can use the JSON event stream mode and a second watchexec instance triggered on signals, or store global state in a file, or use xattrs to store per-file hashes, etc.

**Reply by musjj** [reply, marked as answer]
I ended up using signals instead:

```bash
callback() {
  echo hello world
}

trap callback USR1 USR2
watchexec -- kill -USR1 $$ &
trap "kill $!" EXIT
until wait; do
  :
done
```
watchexec will send a signal to the script itself every time it is triggered. The nice thing here is that you can keep state (e.g. global variables) inbetween runs.
